File: chisel/ui/gtk/filesavechooser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include <gtk/gtk.h>

#include <chisel-native.h>
#include <chisel-native-ui.h>

#include <chisel-native-filesavechooser.h>

native_handle _chisel_native_filesavechooser_create( ) {
	GtkWidget *dialog;
	
	dialog = gtk_file_chooser_dialog_new( "Save File", NULL,
		GTK_FILE_CHOOSER_ACTION_SAVE,
		GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
		GTK_STOCK_SAVE, GTK_RESPONSE_ACCEPT,
		NULL
	);
	
	return dialog;
}

void _chisel_native_filesavechooser_set_show_hidden( native_handle chooser, int flag ) {
	gtk_file_chooser_set_show_hidden( GTK_FILE_CHOOSER(chooser), flag );
}

int _chisel_native_filesavechooser_get_show_hidden( native_handle chooser ) {
	return gtk_file_chooser_get_show_hidden( GTK_FILE_CHOOSER(chooser) );
}

void _chisel_native_filesavechooser_set_can_mkdir( native_handle chooser, int flag ) {
	gtk_file_chooser_set_create_folders( GTK_FILE_CHOOSER(chooser), flag );
}

int _chisel_native_filesavechooser_get_can_mkdir( native_handle chooser ) {
	return gtk_file_chooser_get_create_folders( GTK_FILE_CHOOSER(chooser) );
}

void _chisel_native_filesavechooser_set_allowed_file_types( native_handle chooser, native_handle types ) {
	
}

native_handle _chisel_native_filesavechooser_get_allowed_file_types( native_handle chooser ) {
	return NULL;
}

void _chisel_native_filesavechooser_begin_modal( native_handle chooser, native_handle nwindow ) {
	GtkWidget *dialog = GTK_WIDGET(chooser);
	GtkWidget *window = GTK_WIDGET(nwindow);
	
	assert( dialog != NULL );
	
	int response = gtk_dialog_run( GTK_DIALOG(dialog) );
	
	_chisel_native_filesavechooser_completed_callback( chooser, (response == GTK_RESPONSE_ACCEPT) );
	
	gtk_widget_hide( dialog );
}

native_handle _chisel_native_filesavechooser_get_path( native_handle chooser ) {
	char *filename = gtk_file_chooser_get_filename( GTK_FILE_CHOOSER(chooser) );
	native_handle str = _chisel_native_string_create_with_utf8_bytes( filename, strlen(filename) );
	
	g_free( filename );
	
	return (native_handle)str;
}

File: chisel/ui/macosx/filesavechooser.m
#include <Cocoa/Cocoa.h>
#include <assert.h>

#include <chisel-native.h>
#include <chisel-native-ui.h>

#include <chisel-native-filesavechooser.h>

@interface CLSavePanel : NSSavePanel
// adds prototypes for 10.6 functions
// these messages should only be sent on 10.6 where they are available
- (void)setShowsHiddenFiles:(BOOL)flag;
- (BOOL)showsHiddenFiles;
@end

@interface ChiselSaveDelegate : NSObject
// delegates
- (void)savePanelDidEnd:(NSSavePanel *)sheet returnCode:(int)returnCode contextInfo:(void *)contextInfo;
@end

@implementation ChiselSaveDelegate
- (void)savePanelDidEnd:(NSSavePanel *)sheet returnCode:(int)returnCode contextInfo:(void *)contextInfo {
	[sheet close];
	
	assert( [[sheet URL] isFileURL] );
	_chisel_native_filesavechooser_completed_callback( (native_handle)sheet, (returnCode == NSFileHandlingPanelOKButton) );
}
@end

native_handle _chisel_native_filesavechooser_create( ) {
	NSSavePanel *panel = [NSSavePanel savePanel];
	
	return (native_handle)panel;
}

void _chisel_native_filesavechooser_set_show_hidden( native_handle chooser, int flag ) {
	CLSavePanel *panel = (CLSavePanel *)chooser;
	
	// 10.6 only
	if ( [panel respondsToSelector:@selector(setShowsHiddenFiles:)] ) {
		[panel setShowsHiddenFiles: flag];
	}
}

int _chisel_native_filesavechooser_get_show_hidden( native_handle chooser ) {
	CLSavePanel *panel = (CLSavePanel *)chooser;
	
	// 10.6 only
	if ( [panel respondsToSelector:@selector(showsHiddenFiles:)] ) {
		return [panel showsHiddenFiles];
	} else {
		return 0;
	}
}

void _chisel_native_filesavechooser_set_can_mkdir( native_handle chooser, int flag ) {
	NSSavePanel *panel = (NSSavePanel *)chooser;
	
	[panel setCanCreateDirectories: flag];
}

int _chisel_native_filesavechooser_get_can_mkdir( native_handle chooser ) {
	NSSavePanel *panel = (NSSavePanel *)chooser;
	
	return [panel canCreateDirectories];
}

void _chisel_native_filesavechooser_set_allowed_file_types( native_handle chooser, native_handle types ) {
	NSSavePanel *panel = (NSSavePanel *)chooser;
	NSArray *arrTypes = (NSArray *)types;
	
	[panel setAllowedFileTypes: arrTypes];
}

native_handle _chisel_native_filesavechooser_get_allowed_file_types( native_handle chooser ) {
	NSSavePanel *panel = (NSSavePanel *)chooser;
	
	return (native_handle)[panel allowedFileTypes];
}

void _chisel_native_filesavechooser_begin_modal( native_handle chooser, native_handle nwindow ) {
	NSSavePanel *panel = (NSSavePanel *)chooser;
	NSWindow *window = (NSWindow *)nwindow;
	
	[panel beginSheetForDirectory:nil
		  	file:nil
			modalForWindow:window
			modalDelegate:[[ChiselSaveDelegate alloc] init]
			didEndSelector:@selector(savePanelDidEnd:returnCode:contextInfo:)
			contextInfo:nil
			];
}

native_handle _chisel_native_filesavechooser_get_path( native_handle chooser ) {
	NSSavePanel *panel = (NSSavePanel *)chooser;
	
	return (native_handle)[[panel URL] path];
}
